expand doc snippets to lookups and other plugins

bcoca opened the pull request:
Make snippets work for other plugins

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ansible-doc

##### ADDITIONAL INFO
for lookups:
``` 
ansible-doc -t lookup url -s
```
outputs:
```yaml
# _terms(string): urls to query
# ca_path(string): String of file system path to CA cert bundle to use
# follow_redirects(string): String of urllib2, all/yes, safe, none to determine how redirects are followed, see RedirectHandlerFactory for more information
# force(boolean): Whether or not to set "cache-control" header with value "no-cache"
# force_basic_auth(boolean): Force basic authentication
# headers(dictionary): HTTP request headers
# http_agent(string): User-Agent to use in the request. The default was changed in 2.11 to C(ansible-httpget).
# password(string): Password to use for HTTP authentication.
# split_lines(boolean): Flag to control if content is returned as a list of lines or as a single text blob
# timeout(float): How long to wait for the server to send data before giving up
# unix_socket(string): String of file system path to unix socket file to use when establishing connection to the provided url
# unredirected_headers(list): A list of headers to not attach on a redirected request
# use_gssapi(boolean): ['Use GSSAPI handler of requests', 'As of Ansible 2.11, GSSAPI credentials can be specified with I(username) and I(password).']
# use_proxy(boolean): Flag to control if the lookup will observe HTTP proxy environment variables when present.
# username(string): Username to use for HTTP authentication.
# validate_certs(boolean): Flag to control SSL certificate validation

lookup('url', < _terms >, ca_path='None', follow_redirects='urllib2', force=False, force_basic_auth=False, headers={}, http_agent='ansible-httpget', password='None', split_lines=True, timeout=10, unix_socket='None', unredirected_headers=None, use_gssapi=False, use_proxy=True, username='None', validate_certs=True)
```

for inventory
```
 ansible-doc -s -t inventory community.general.nmap
```
outputs:
```yaml
# nmap:
address: (required) # Network IP or range of IPs to scan, you can use a simple range (10.2.2.15-25) or CIDR notation.
cache: False     # Toggle to enable/disable the caching of the inventory's source data, requires a cache plugin setup to work.
cache_connection: None      # Cache connection data or path, read cache plugin documentation for specifics.
cache_plugin: memory    # Cache plugin to use for the inventory's source data.
cache_prefix: ansible_inventory_ # Prefix to use for cache plugin files/tables
cache_timeout: 3600      # Cache duration in seconds
compose: {}        # Create vars from jinja2 expressions.
exclude: None      # list of addresses to exclude
groups: {}        # Add hosts to group based on Jinja2 conditionals.
ipv4: True      # use IPv4 type addresses
ipv6: True      # use IPv6 type addresses
keyed_groups: []        # Add hosts to group based on the values of a variable.
leading_separator: True      # Use in conjunction with keyed_groups. By default, a keyed group that does not have a prefix or a separator provided will have a name that starts with an
                             # underscore. This is because the default prefix is "" and the default separator is "_". Set this option to False to omit
                             # the leading underscore (or other separator) if no prefix is given. If the group name is derived from a mapping the [...]
plugin: (required) # token that ensures this is a source file for the 'nmap' plugin.
ports: True      # Enable/disable scanning for open ports
strict: False     # If `yes' make invalid entries a fatal error, otherwise skip and continue. Since it is possible to use facts in the expressions they might not always be
                             # available and we ignore those errors by default.
use_extra_vars: False     # Merge extra vars into the available variables for composition (highest precedence).
```

bcoca committed:
beetah snippets

bcoca committed:
better str

bcoca committed:
comment on terms

bcoca committed:
max lines added to ini/non module yaml output

bcoca committed:
pruttier ini

bcoca committed:
more visible requried

bcoca committed:
itsacomment

bcoca committed:
spaced

bcoca committed:
whitey

bcoca committed:
simpler handling

Shrews commented on lib/ansible/cli/doc.py:
```diff
@@ -1197,3 +1195,120 @@ def get_man_text(doc, collection_name='', plugin_type=''):
+        if required:
+            default = '<REQUIRED>'
+        else:
+            default = opt.get('default', 'None')
```
`'None'` or `None` here?

I'd argue if it isn't required, and there is no default, then not displaying it at all is the better option.

bcoca replied:
using string to show None, which is the default when no default is specified

bcoca committed:
no need for comment title

webknjaz commented on lib/ansible/cli/doc.py:
```diff
@@ -400,32 +400,42 @@ def init_parser(self):
 
+        # modifiers
         exclusive = self.parser.add_mutually_exclusive_group()
+        # TODO: wanr if not used with -t roles
```
```suggestion
        # TODO: warn if not used with -t roles
```

webknjaz commented on lib/ansible/cli/doc.py:
```diff
@@ -948,30 +958,20 @@ def print_paths(finder):
 
     @staticmethod
-    def get_snippet_text(doc):
+    def get_snippet_text(ptype, doc):
```
Nit: it's best to add new args at the end for compat

webknjaz commented on lib/ansible/cli/doc.py:
```diff
@@ -1251,3 +1251,118 @@ def get_man_text(doc, collection_name='', plugin_type=''):
+        text.append("# %s:" % doc.get('plugin', doc.get('name')))
+
+    pad = 29
+    subdent = " " * pad + '# '
```
How about 
```suggestion
    subdent = '# '.rjust(31)
```

bcoca committed:
Update lib/ansible/cli/doc.py

Co-authored-by: Sviatoslav Sydorenko <wk.cvs.github@sydorenko.org.ua>

bcoca committed:
feedback incorporated

abadger commented on lib/ansible/cli/doc.py:
```diff
@@ -948,30 +958,20 @@ def print_paths(finder):
+            _do_yaml_snippet(text, doc)
+        elif 'options' in doc:
+            _do_ini_snippet(text, doc)
+        else:
```
This should probably be an error condition (or print the man text as the previous version did but an error feels like it's more informative).

bcoca replied:
Not sure it warrants an error, a warning perhaps?

This can be the result of the specific plugin having no options, as used to be the case for script inventory.

abadger commented on lib/ansible/cli/doc.py:
```diff
@@ -948,30 +958,20 @@ def print_paths(finder):
 
     @staticmethod
-    def get_snippet_text(doc):
+    def get_snippet_text(doc, pytpe='module'):
```
```suggestion
    def get_snippet_text(doc, ptype='module'):
```

abadger commented on lib/ansible/cli/doc.py:
```diff
@@ -829,8 +839,8 @@ def format_plugin_doc(plugin, plugin_type, doc, plainexamples, returndocs, metad
-        if context.CLIARGS['show_snippet'] and plugin_type == 'module':
-            text = DocCLI.get_snippet_text(doc)
+        if context.CLIARGS['show_snippet']:
+            text = DocCLI.get_snippet_text(doc, plugin_type)
```
Now that this no longer depends on plugin_type to activate the snippet path, this condition should be moved to the caller.  It's a decision about the overall format the data should be displayed as rather than a nitty-gritty decision about how to display a specific piece of data in this particular format. It's the roughly the same level of decision as whether the output should be formatted as a json string or as human readable output which is being done in the caller.

bcoca replied:
It is something I have been thinking about for a while, making a greater separation between data sourcing and formatting, it will probably simplify things a lot, but that might be worth it's own PR.

bcoca committed:
Update lib/ansible/cli/doc.py

Co-authored-by: Toshio Kuratomi <a.badger@gmail.com>

bcoca committed:
added env var drop

bcoca committed:
lookup snip now has comments

bcoca committed:
simplified to just deal with 'play' insertable

  lookups are added to existing modules
  filters and tests can be added once documentable

samdoran commented:
!waffling core_review

bcoca committed:
made py2 compatible

bcoca committed:
whiteeeey!

bcoca committed:
also generates a sample inventory config

bcoca committed:
made exceptions for the exceptional plugins

bcoca committed:
clog
